### PR TITLE
feat: dvm 2.0 upgrade improvement

### PR DESCRIPTION
**Pull request by md0x**
Signed-off-by: Pablo Maldonado <pablo@umaproject.org>

**Motivation**

The previous migration scripts were incomplete, in this PR we fix and add the missing steps and the possibility of doing a downgrade after an upgrade to implicitly test the contracts functionalities. 

When running the scripts, once an upgrade is completed, it is possible to perform multiple upgrades and downgrades by copying and executing the commands provided by the CLI.

**Summary**

- Fix/Add missing steps in the upgrade process
- Add more control to be able to do an optional downgrade after performing and upgrade
- Add simulation script to simulate a full voting cycle and execution of transactions in VotingV2

**Testing**

Check a box to describe how you tested these changes and list the steps for reviewers to test.

- [ ]  Ran end-to-end test, running the code as in production
- [ ]  New unit tests created
- [ ]  Existing tests adequate, no new tests required
- [x]  All existing tests pass
- [ ]  Untested

### CODE REVIEW

File: packages/core/contracts/umips/VotingUpgraderV2.sol [hunk 1]
```diff
@@ -19,7 +19,6 @@ struct OwnableContracts {
     Ownable governorHub;
     Ownable bobaParentMessenger;
     Ownable optimismParentMessenger;
-    Ownable proposer;
```
Comment (md0x) [inline packages/core/contracts/umips/VotingUpgraderV2.sol:22]: Proposer has been moved to a separate argument

File: packages/scripts/src/upgrade-tests/voting2/2_Verify.ts [hunk 2]
```diff
@@ -34,50 +43,53 @@ async function main() {
...
 
   const finder = await getContractInstance<FinderEthers>("Finder");
-  const governor = await getContractInstance<GovernorEthers>("Governor");
-  const oldVoting = await getContractInstance<VotingEthers>("Voting");
+  const registry = await getContractInstance<RegistryEthers>("Registry");
   const votingToken = await getContractInstance<VotingTokenEthers>("VotingToken");
 
-  const votingV2Factory: VotingV2Ethers__factory = await getContractFactory("VotingV2");
-  const votingV2 = await votingV2Factory.attach(votingV2Address);
+  const governor = await getContractInstance<GovernorEthers>("Governor", process.env[OLD_CONTRACTS.governor]);
```
Comment (md0x) [inline packages/scripts/src/upgrade-tests/voting2/2_Verify.ts:52]: If the env var is not set, `getContractInstance` defaults to the deployed address stored in the repo

File: packages/scripts/src/upgrade-tests/voting2/2_Verify.ts [hunk 3]
```diff
@@ -86,15 +98,54 @@ async function main() {
...
   console.log("✅ Voting token owner role correctly set!");
 
-  console.log("Verified!");
+  console.log(" 6. Governor v2 is the owner of proposer...");
+  assert.equal((await proposer.owner()).toLowerCase(), governorV2.address.toLowerCase());
+  console.log("✅ Proposer owner role correctly set!");
+
+  console.log(" 7. Governor v2 is the owner of proposer v2...");
+  assert.equal((await proposerV2.owner()).toLowerCase(), governorV2.address.toLowerCase());
+  console.log("✅ Proposer v2 owner role correctly set!");
```
Comment (md0x) [inline packages/scripts/src/upgrade-tests/voting2/2_Verify.ts:115]: Both proposers ownerships are transferred to the GovernorV2

File: packages/scripts/src/upgrade-tests/voting2/2_Verify.ts [hunk 4]
```diff
@@ -86,15 +98,54 @@ async function main() {
...
+  console.log(" 8. Governor v2 is registered in the registry...");
+  assert(await registry.isContractRegistered(governorV2.address));
+  console.log("✅ Governor v2 registered in registry!");
+
+  console.log(" 9. Proposer v2 is registered in the regstry...");
+  assert(await registry.isContractRegistered(proposerV2.address));
+  console.log("✅ Proposer v2 registered in registry!");
+
+  console.log(" 10. Governor v2 received all the voting tokens from Governor...");
+  assert((await votingToken.balanceOf(governorV2.address)).gt(hre.web3.utils.toWei("30000000", "ether")));
```
Comment (md0x) [inline packages/scripts/src/upgrade-tests/voting2/2_Verify.ts:126]: Maybe this hardcoded value is unnecessary and we should just check that the balance is >0

**Comment by chrismaree** [comment]
I re-based this to fix a CI issue

File: packages/core/contracts/umips/VotingUpgraderV2.sol [hunk 5]
```diff
@@ -73,6 +76,7 @@ contract VotingUpgraderV2 {
         address _newGovernor,
         address _existingVoting,
         address _newVoting,
+        address _existingProposer,
```
Comment (chrismaree) [inline packages/core/contracts/umips/VotingUpgraderV2.sol:79]: why is this needed?

File: packages/scripts/src/admin-proposals/simulateVoteV2.ts [hunk 6]
```diff
@@ -0,0 +1,246 @@
+// Description:
+// - Simulate voting affirmatively on any pending Admin Proposals
+
+import { advanceBlockAndSetTime, computeVoteHashAncillary, getRandomSignedInt, getWeb3ByChainId } from "@uma/common";
+import { GovernorV2Ethers, VotingTokenEthers, VotingV2Ethers } from "@uma/contracts-node";
+import { BigNumberish, BytesLike } from "ethers";
+// import { REQUIRED_SIGNER_ADDRESSES, SECONDS_PER_DAY, YES_VOTE } from "../utils/constants";
```
Comment (chrismaree) [inline packages/scripts/src/admin-proposals/simulateVoteV2.ts:7]: remove this?

File: packages/scripts/src/admin-proposals/simulateVoteV2.ts [hunk 7]
```diff
@@ -0,0 +1,246 @@
...
+import { getContractInstance } from "../utils/contracts";
+
+const SECONDS_PER_DAY = 86400;
+const YES_VOTE = "1";
+const REQUIRED_SIGNER_ADDRESSES = {
+  deployer: "0x2bAaA41d155ad8a4126184950B31F50A1513cE25",
+  foundation: "0x7a3A1c2De64f20EB5e916F40D11B01C441b2A8Dc",
+  account_with_uma: "0xcb287f69707d84cbd56ab2e7a4f32390fa98120b",
```
Comment (chrismaree) [inline packages/scripts/src/admin-proposals/simulateVoteV2.ts:15]: this is not used.

File: packages/scripts/src/admin-proposals/simulateVoteV2.ts [hunk 8]
```diff
@@ -0,0 +1,246 @@
...
+import { advanceBlockAndSetTime, computeVoteHashAncillary, getRandomSignedInt, getWeb3ByChainId } from "@uma/common";
+import { GovernorV2Ethers, VotingTokenEthers, VotingV2Ethers } from "@uma/contracts-node";
+import { BigNumberish, BytesLike } from "ethers";
+// import { REQUIRED_SIGNER_ADDRESSES, SECONDS_PER_DAY, YES_VOTE } from "../utils/constants";
+import { getContractInstance } from "../utils/contracts";
+
+const SECONDS_PER_DAY = 86400;
+const YES_VOTE = "1";
+const REQUIRED_SIGNER_ADDRESSES = {
+  deployer: "0x2bAaA41d155ad8a4126184950B31F50A1513cE25",
```
Comment (chrismaree) [inline packages/scripts/src/admin-proposals/simulateVoteV2.ts:13]: this is also not used

File: packages/scripts/src/admin-proposals/simulateVoteV2.ts [hunk 9]
```diff
@@ -0,0 +1,246 @@
...
+//   to simulate what would happen if the proposal were to pass and to verify that contract state changes as expected.
+// - Vote Simulate: NODE_URL_1=http://localhost:9545 node ./packages/scripts/src/admin-proposals/simulateVoteV2.js --network mainnet-fork
+
+const hre = require("hardhat");
+const { toWei } = hre.web3.utils;
+require("dotenv").config();
+const assert = require("assert").strict;
+
+async function simulateVoteV2() {
+  const foundationSigner = await hre.ethers.getSigner(REQUIRED_SIGNER_ADDRESSES["foundation"]);
```
Comment (chrismaree) [inline packages/scripts/src/admin-proposals/simulateVoteV2.ts:31]: it might be cleaner to replace the `REQUIRED_SIGNER_ADDRESSES["foundation"]` with just `FOUNDATION_WALLET` seeing it's used all over the place and you dont use the other wallets imported in the same object.

File: packages/scripts/src/admin-proposals/simulateVoteV2.ts [hunk 10]
```diff
@@ -0,0 +1,246 @@
...
+
+  const governorV2Address = process.env["GOVERNOR_V2_ADDRESS"];
+  const votingV2Address = process.env["VOTING_V2_ADDRESS"];
+
+  const governorV2 = await getContractInstance<GovernorV2Ethers>("GovernorV2", governorV2Address);
+  const votingV2 = await getContractInstance<VotingV2Ethers>("VotingV2", votingV2Address);
+  const votingToken = await getContractInstance<VotingTokenEthers>("VotingToken");
+
+  const web3 = getWeb3ByChainId(1);
+  const accounts = await web3.eth.getAccounts();
```
Comment (chrismaree) [inline packages/scripts/src/admin-proposals/simulateVoteV2.ts:41]: this is a little bit janky. we are using _both_ web3 and ethers in this script. it would be better to just use one or the other, ideally. I think you might need to move to just using web3, unfortunately, seeing some methods, like `advanceBlockAndSetTime` are only configured to work with web3. else, you'll need to make this work with ethers as well.

File: packages/scripts/src/upgrade-tests/voting2/0_Deploy.ts [hunk 11]
```diff
@@ -82,14 +85,44 @@ async function main() {
...
+  ☝️ PROPOSAL: Run the following command to propose the migration to VotingV2, GovernorV2 and ProposerV2:
+  ${VOTING_UPGRADER_ADDRESS}=${votingUpgrader.address} \\
+  ${NEW_CONTRACTS.voting}=${votingV2.address} \\
+  ${NEW_CONTRACTS.governor}=${governorV2.address} \\
+  ${NEW_CONTRACTS.proposer}=${proposerV2.address} \\
+  yarn hardhat run ./src/upgrade-tests/voting2/1_Propose.ts --network ${hre.network.name}`.replace(/  +/g, "")
+  );
 }
 
 main().then(
```
Comment (chrismaree) [inline packages/scripts/src/upgrade-tests/voting2/0_Deploy.ts:128]: you'll extend this in a following PR for the additional functionality for the needed contracts, right?

Comment (md0x) [reply packages/scripts/src/upgrade-tests/voting2/0_Deploy.ts:128]: Yes, that was my plan!

File: packages/scripts/src/admin-proposals/simulateVoteV2.ts (hunk 10 above)
Comment (md0x) [reply packages/scripts/src/admin-proposals/simulateVoteV2.ts]: [Fixed](https://github.com/UMAprotocol/protocol/pull/4190/commits/7fd6601f32da154b095fe49af95329d1ed943b56)

File: packages/scripts/src/admin-proposals/simulateVoteV2.ts [hunk 12]
```diff
@@ -0,0 +1,241 @@
...
+
+require("dotenv").config();
+const assert = require("assert").strict;
+
+const SECONDS_PER_DAY = 86400;
+const YES_VOTE = "1";
+
+const FOUNDATION_WALLET = "0x7a3A1c2De64f20EB5e916F40D11B01C441b2A8Dc";
+
+const increaseEvmTime = async (time: number) => {
```
Comment (chrismaree) [inline packages/scripts/src/admin-proposals/simulateVoteV2.ts:19]: nice, this is useful. we can work on adding this to a common util at some point in the future.

File: packages/scripts/src/upgrade-tests/voting2/1_Propose.ts [hunk 13]
```diff
@@ -1,153 +1,270 @@
...
+      if (!registerProposerV2Tx.data) throw new Error("registerProposerV2Tx.data is empty");
+      adminProposalTransactions.push({ to: registry.address, value: 0, data: registerProposerV2Tx.data });
+      console.log("4.f.2 Register the ProposerV2 as a verified contract", registerProposerV2Tx.data);
+    }
+
+    if (!governorV2Registered) {
+      const registerGovernorV2Tx = await registry.populateTransaction.registerContract([], governorV2.address);
+      if (!registerGovernorV2Tx.data) throw new Error("registerGovernorV2Tx.data is empty");
+      adminProposalTransactions.push({ to: registry.address, value: 0, data: registerGovernorV2Tx.data });
+      console.log("4.f.3 Register the ProposerV2 as a verified contract", registerGovernorV2Tx.data);
```
Comment (Reinis-FRP) [inline packages/scripts/src/upgrade-tests/voting2/1_Propose.ts:161]:
```suggestion
      console.log("4.f.3 Register the GovernorV2 as a verified contract", registerGovernorV2Tx.data);
```

Comment (md0x) [reply packages/scripts/src/upgrade-tests/voting2/1_Propose.ts]: Thanks!
